File: frontend/src/lib/api/notes.ts
import { API_BASE_URL, apiFetch, apiFetchBlob } from "@/lib/api";
import { getToken } from "@/lib/auth";

export type Note = {
  id: string;
  user_id: string;
  author_name: string;
  title: string;
  description: string | null;
  course_code: string;
  semester: string;
  file_size_bytes: number;
  created_at: string;
};

export type NoteListResponse = {
  items: Note[];
  total: number;
};

export type NoteFilters = {
  course_code?: string;
  semester?: string;
  q?: string;
  limit?: number;
  offset?: number;
};

function buildQuery(filters: NoteFilters): string {
  const params = new URLSearchParams();
  if (filters.course_code) params.set("course_code", filters.course_code);
  if (filters.semester) params.set("semester", filters.semester);
  if (filters.q) params.set("q", filters.q);
  if (filters.limit !== undefined) params.set("limit", String(filters.limit));
  if (filters.offset !== undefined) params.set("offset", String(filters.offset));
  const query = params.toString();
  return query ? `?${query}` : "";
}

export async function listNotes(filters: NoteFilters = {}): Promise<NoteListResponse> {
  return apiFetch<NoteListResponse>(`/notes${buildQuery(filters)}`, {
    method: "GET",
    token: getToken(),
  });
}

export async function getNote(id: string): Promise<Note> {
  return apiFetch<Note>(`/notes/${id}`, {
    method: "GET",
    token: getToken(),
  });
}

export async function uploadNote(form: FormData): Promise<Note> {
  return apiFetch<Note>("/notes", {
    method: "POST",
    body: form,
    token: getToken(),
  });
}

export async function deleteNote(id: string): Promise<void> {
  await apiFetch<null>(`/notes/${id}`, {
    method: "DELETE",
    token: getToken(),
  });
}

export async function downloadNote(note: Note): Promise<void> {
  const blob = await apiFetchBlob(`/notes/${note.id}/download`, getToken());
  const url = URL.createObjectURL(blob);
  try {
    const a = document.createElement("a");
    a.href = url;
    const safeTitle = note.title.replace(/[\\/:*?"<>|]/g, "").trim() || "note";
    a.download = `${safeTitle}.pdf`;
    document.body.appendChild(a);
    a.click();
    a.remove();
  } finally {
    setTimeout(() => URL.revokeObjectURL(url), 0);
  }
}

export const noteDownloadUrl = (id: string) => `${API_BASE_URL}/notes/${id}/download`;

File: backend/app/routes/notes.py
from typing import Annotated
from uuid import UUID

from fastapi import APIRouter, File, Form, Query, UploadFile
from fastapi.responses import FileResponse, Response

from app.core.deps import CurrentUserId, DbSession
from app.repositories.notes import NoteListItem
from app.schemas.notes import NoteListResponse, NoteRead
from app.services.notes import NoteService

router = APIRouter(prefix="/notes", tags=["notes"])


def _to_read(item: NoteListItem) -> NoteRead:
    return NoteRead(
        id=item.note.id,
        user_id=item.note.user_id,
        author_name=item.author_name,
        title=item.note.title,
        description=item.note.description,
        course_code=item.note.course_code,
        semester=item.note.semester,
        file_size_bytes=item.note.file_size_bytes,
        created_at=item.note.created_at,
    )


@router.get("", response_model=NoteListResponse)
def list_notes(
    db: DbSession,
    _user_id: CurrentUserId,
    course_code: Annotated[str | None, Query()] = None,
    semester: Annotated[str | None, Query()] = None,
    q: Annotated[str | None, Query()] = None,
    limit: Annotated[int, Query(ge=1, le=100)] = 20,
    offset: Annotated[int, Query(ge=0)] = 0,
) -> NoteListResponse:
    result = NoteService(db).list(
        course_code=course_code,
        semester=semester,
        q=q,
        limit=limit,
        offset=offset,
    )
    return NoteListResponse(
        items=[_to_read(item) for item in result.items],
        total=result.total,
    )


@router.post("", response_model=NoteRead, status_code=201)
def create_note(
    db: DbSession,
    user_id: CurrentUserId,
    title: Annotated[str, Form(min_length=1, max_length=200)],
    course_code: Annotated[str, Form(min_length=1, max_length=50)],
    semester: Annotated[str, Form(min_length=1, max_length=20)],
    file: Annotated[UploadFile, File()],
    description: Annotated[str | None, Form(max_length=2000)] = None,
) -> NoteRead:
    item = NoteService(db).create(
        user_id=user_id,
        title=title,
        description=description,
        course_code=course_code,
        semester=semester,
        file=file,
    )
    return _to_read(item)


@router.get("/{note_id}", response_model=NoteRead)
def get_note(note_id: UUID, db: DbSession, _user_id: CurrentUserId) -> NoteRead:
    return _to_read(NoteService(db).get(note_id))


@router.get("/{note_id}/download")
def download_note(note_id: UUID, db: DbSession, _user_id: CurrentUserId) -> FileResponse:
    note, path = NoteService(db).file_for_download(note_id)
    safe_title = note.title.replace('"', "").strip() or "note"
    filename = f"{safe_title}.pdf"
    return FileResponse(
        path=path,
        media_type="application/pdf",
        filename=filename,
    )


@router.delete("/{note_id}", status_code=204)
def delete_note(note_id: UUID, db: DbSession, user_id: CurrentUserId) -> Response:
    NoteService(db).delete(note_id=note_id, requester_id=user_id)
    return Response(status_code=204)
